Kernel/VirtIO/VirtIOConsolePort.cpp

```diff
@@ -0,0 +1,167 @@
+/*
+ * Copyright (c) 2021, the SerenityOS developers.
+ * Copyright (c) 2021, Kyle Pereira <hey@xylepereira.me>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <Kernel/VirtIO/VirtIOConsole.h>
+#include <Kernel/VirtIO/VirtIOConsolePort.h>
+
+namespace Kernel {
+
+unsigned VirtIOConsolePort::next_device_id = 0;
+
+VirtIOConsolePort::VirtIOConsolePort(unsigned port, VirtIOConsole& console)
+    : CharacterDevice(229, next_device_id++)
+    , m_console(console)
+    , m_port(port)
+{
+    m_receive_buffer = make<RingBuffer>("VirtIOConsolePort Receive", RINGBUFFER_SIZE);
+    m_transmit_buffer = make<RingBuffer>("VirtIOConsolePort Transmit", RINGBUFFER_SIZE);
+    m_receive_queue = m_port == 0 ? 0 : m_port * 2 + 2;
+    m_transmit_queue = m_port == 0 ? 1 : m_port * 2 + 3;
+    init_receive_buffer();
+}
+
+void VirtIOConsolePort::init_receive_buffer()
+{
+    auto& queue = m_console.get_queue(m_receive_queue);
+    ScopedSpinLock queue_lock(queue.lock());
+    VirtIOQueueChain chain(queue);
+
+    auto buffer_start = m_receive_buffer->start_of_region();
+    auto did_add_buffer = chain.add_buffer_to_chain(buffer_start, RINGBUFFER_SIZE, BufferType::DeviceWritable);
+    VERIFY(did_add_buffer);
+    m_console.supply_chain_and_notify(m_receive_queue, chain);
+}
+
+void VirtIOConsolePort::handle_queue_update(Badge<VirtIOConsole>, u16 queue_index)
+{
+    dbgln_if(VIRTIO_DEBUG, "VirtIOConsolePort: Handle queue update for port {}", m_port);
+    VERIFY(queue_index == m_transmit_queue || queue_index == m_receive_queue);
+    if (queue_index == m_receive_queue) {
+        auto& queue = m_console.get_queue(m_receive_queue);
+        ScopedSpinLock queue_lock(queue.lock());
+        size_t used;
+        VirtIOQueueChain popped_chain = queue.pop_used_buffer_chain(used);
+
+        ScopedSpinLock ringbuffer_lock(m_receive_buffer->lock());
+        auto used_space = m_receive_buffer->reserve_space(used).value();
+        auto remaining_space = m_receive_buffer->bytes_till_end();
+
+        // Our algorithm always has only one buffer in the queue.
+        VERIFY(popped_chain.length() == 1);
+        size_t unused;
+        VERIFY(queue.pop_used_buffer_chain(unused).is_empty());
```

Comment:
This should still be changed to:
```suggestion
        VERIFY(!queue.new_data_available());
```